### ISSUE TITLE
Numpy invalid division issue in restored model

### ISSUE THREAD

**Issue by amir-zeldes** [riga/tfdeploy#24]
Hi, I'm using a multilayer perceptron with dropout and exponential decay, adapted from the MNIST tutorial in tensorflow. I'm interested in deploying a softmax layer to get class membership probabilities rather than hard predictions. I can train and test the model successfully in tensorflow like this, but when I deploy and restore the model, I run into an "invalid value encountered in true_divide" error from numpy. Here is the (abbreviated) training part:

```
...
# tf Graph input
x = tf.placeholder("float", shape=[None, n_input],name="input")
y = tf.placeholder("float", [None, n_classes])

# Dropout
keep_prob = tf.placeholder(tf.float32, name="keep_prob")
dropout = 0.75
...

def multilayer_perceptron(x, weights, biases, dropout):
	# Hidden layer with RELU activation
	layer_1 = tf.add(tf.matmul(x, weights['h1']), biases['b1'])
	layer_1 = tf.nn.relu(layer_1)
	# Apply dropout
	layer_1 = tf.nn.dropout(layer_1, dropout)

	# Hidden layer with RELU activation
	layer_2 = tf.add(tf.matmul(layer_1, weights['h2']), biases['b2'])
	layer_2 = tf.nn.relu(layer_2)
	# Output layer with linear activation
	out_layer = tf.matmul(layer_2, weights['out']) + biases['out']
	return out_layer
...
pred = multilayer_perceptron(x, weights, biases, keep_prob)
...
# Training cycle
for epoch in range(training_epochs):
	avg_cost = 0.
	total_batch = int(num_examples/batch_size)
	# Loop over all batches
	for i in range(total_batch):
		batch_x, batch_y = next_batch(data,batch_size)
		# Run optimization op (backprop) and cost op (to get loss value)
		_, c = sess.run([optimizer, cost], feed_dict={x: batch_x, y: batch_y, keep_prob: dropout})
		# Compute average loss
		avg_cost += c / total_batch
	# Display logs per epoch step
	if epoch % display_step == 0:
		print("Epoch:", '%04d' % (epoch+1), "cost=", "{:.9f}".format(avg_cost))
print("Optimization Finished!")

y = tf.nn.softmax(pred, name="output")

# Evaluate some row in test data, without dropout, this seems to work
print(y.eval({x: test_x[5, :].reshape(1, 262), keep_prob: 1.}))

# Now save tfdeploy model to disk
model = tfdeploy.Model()
model.add(y, sess) # y and all its ops and related tensors are added recursively
model.save("tdeploy_model.pkl")
```

This first part works fine. But now the problem comes up:
```
nn = tfdeploy.Model("tdeploy_model.pkl")
x, y, keep_prob = self.nn.get("input", "output", "keep_prob")
data = np.loadtxt("some_data.tab", delimiter="\t")
# Pick some row
data = data[4:5,0:data.shape[1]-2]
# Get a prediction
y.eval({x: data, keep_prob: 1.})
```

This last operation throws a warning as below:

```
tfdeploy.py:2098: RuntimeWarning: invalid value encountered in true_divide
  return np.divide(e, np.sum(e, axis=-1, keepdims=True)),
```

And the predicted probabilities are `[nan]`, instead of the floats I see when running directly under tensorflow. Any ideas on what is happening here? Thanks for your help!

**Comment by amir-zeldes** [comment]
Quick update: I was able to resolve this using some variable scaling - numbers are now less close to 1 and 0 and the error doesn't trigger. I'm still not sure why I was seeing different behavior in tf and tfdeploy.

**Comment by riga** [comment]
Sorry for the long delay.

Maybe this was caused some numerical instabilities due to large exp values and floating points. Do you still experience any issues with that?

**Comment by amir-zeldes** [comment]
Yes, I also suspected it had something to do with rounding losses. As I said above, rescaling some variables got rid of the error, so I don't have an acute problem with this right now. Thanks!

**Comment by riga** [comment]
Ok, cool. Thanks for the response!
